circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/SaveableStateHolder.kt

```diff
@@ -0,0 +1,117 @@
+// Copyright (C) 2020 Slack Technologies, LLC
+// SPDX-License-Identifier: Apache-2.0
+package com.slack.circuit.foundation
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.saveable.LocalSaveableStateRegistry
+import androidx.compose.runtime.saveable.SaveableStateRegistry
+import androidx.compose.runtime.saveable.Saver
+import androidx.compose.runtime.saveable.rememberSaveable
+
+/**
+ * Allows to save the state defined with [rememberSaveable] for the subtree before disposing it to
+ * make it possible to compose it back next time with the restored state. It allows different
+ * navigation patterns to keep the ui state like scroll position for the currently not composed
+ * screens from the backstack.
+ *
+ * The content should be composed using [SaveableStateProvider] while providing a key representing
+ * this content. Next time [SaveableStateProvider] will be used with the same key its state will be
+ * restored.
+ */
+internal interface SaveableStateHolder {
+  /**
+   * Put your content associated with a [key] inside the [content]. This will automatically save all
+   * the states defined with [rememberSaveable] before disposing the content and will restore the
+   * states when you compose with this key again.
+   *
+   * @param key to be used for saving and restoring the states for the subtree. Note that on Android
+   *   you can only use types which can be stored inside the Bundle.
+   */
+  @Composable fun <R> SaveableStateProvider(key: Any, content: @Composable () -> R): R
+
+  /** Removes the saved state associated with the passed [key]. */
+  fun removeState(key: Any)
+}
+
+/** Creates and remembers the instance of [SaveableStateHolder]. */
+@Composable
+internal fun rememberSaveableStateHolderWithReturn(): SaveableStateHolder =
+  rememberSaveable(saver = SaveableStateHolderImpl.Saver) { SaveableStateHolderImpl() }
+    .apply { parentSaveableStateRegistry = LocalSaveableStateRegistry.current }
+
+private class SaveableStateHolderImpl(
+  private val savedStates: MutableMap<Any, Map<String, List<Any?>>> = mutableMapOf()
+) : SaveableStateHolder {
+  private val registryHolders = mutableMapOf<Any, RegistryHolder>()
+  var parentSaveableStateRegistry: SaveableStateRegistry? = null
+
+  @Composable
+  override fun <R> SaveableStateProvider(key: Any, content: @Composable () -> R): R {
+    val registryHolder = remember {
```

Comment:
This was using `ReusableContent` but I couldn't get it to work (even copying it in and doing similar to `withCompositionLocalProvider`). Everything works without it 🤷